lib/spack/spack/mirror.py

```diff
@@ -60,17 +61,27 @@ def mirror_archive_filename(spec, fetcher, resourceId=None):
                 raise MirrorError(
                     "%s version does not specify an extension" % spec.name +
                     " and could not parse extension from %s" % fetcher.url)
+            # If the archive name is different to the package name
+            if spec.package.name != url.parse_name(fetcher.url):
+                name = url.parse_name(fetcher.url)
```

Comment:
Out of curiosity, why change this method? It chooses the name for the file to be stored in mirrors and the cache. Do you think it is more straightforward to pick a name based on the archive name vs. the package name? I figure either way is equally clear since this function assumes a name format to begin with (and that name format still doesn't match all archive naming schemes).

I presume that nothing broke prior to your changes. If something did break, or if you think this is a better approach, IMO it could be simplified by having an unconditional ```name = url.parse_name```, set ```name = package.name``` in the initial declaration, and simplifying the below setting of ```filename```.